evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -258,6 +238,120 @@ pub(crate) fn generate_segment<F: Field>(
     )))
 }
 
+pub(crate) fn generate_segment_given_data<F: Field>(
```

Comment:
If the generation loop is inside the function instead of outside (currently conditioned by the return value of this function), I think we can avoid one clone of the memory per segment.
Here, after a segment, you have to clone the memory after to create the new batch of data, then at the next execution you need to clone the same data to initialize the interpreter.